### PR TITLE
Adding the Service Logs Streaming feature

### LINKED ISSUE #474: Implement a solution for streaming pod logs
This requirement is the result of experiments made in order to stream pod logs directly to the users terminal, without requiring for them to access the cluster.

FYI  @mchoma

### PR DESCRIPTION
**Pull request by fabiobrz**
Fix #474 

CC @mchoma, @simkam, @tommaso-borgato

Please make sure your PR meets the following requirements:
- [x] Pull Request contains a description of the changes
- [x] Pull Request does not include fixes for multiple issues/topics
- [x] Code is formatted, imports ordered, code compiles and tests are passing
- [x] Code is self-descriptive and/or documented

### CODE REVIEW

**Comment by mchoma** [comment]
I am looking forward to use this feature :)

**Comment by mchoma** [comment]
I could not resist to use it immediatelly :)  and it works!!! Just colouring does not work in my case, but could be just me.

**Comment by fabiobrz** [comment]
> I could not resist to use it immediatelly :) and it works!!! Just colouring does not work in my case, but could be just me.

Thanks for your feedback @mchoma ! I'd like to check about the coloring thing, how did you enable the feature? e.g.: did you pass `xtf.sls.enabled=true`? This is for me to understand whether logs are sent to STDOUT or to a file.

Actually this wouldn't change a thing because I've been using `cat` to see a generated log file and I could see colors there as well, but I'd focus on your use case to double check.

**Comment by mnovak1** [comment]
@fabiobrz I like this idea. This is something we miss and will be useful for everyone. Thanks for raising PR. I'll review it as soon as possible :-)

File: junit5/src/main/resources/META-INF/services/org.junit.jupiter.api.extension.Extension [hunk 1]
```diff
@@ -0,0 +1 @@
+cz.xtf.junit5.extensions.ServiceLogsStreamingRunner
```
Comment (simkam) [inline junit5/src/main/resources/META-INF/services/org.junit.jupiter.api.extension.Extension:1]: other extensions aren't enabled by default in services, why this should be an exception?

**Review by simkam — CHANGES_REQUESTED** [review]
Could you please review the visibility of classes? It looks like all are public but are only used by extension within the package, perhaps they can be changed to package-private? Let's expose as minimum as possible.

File: junit5/src/main/resources/META-INF/services/org.junit.jupiter.api.extension.Extension (hunk 1 above)
Comment (fabiobrz) [reply junit5/src/main/resources/META-INF/services/org.junit.jupiter.api.extension.Extension:1]: Thanks @simkam - This was the solution that came to my mind when I though about how to enable the feature. 
Per se, it had a requirement allowing for it to be started either via annotation (on individual classes for developer UX) or via system properties (automation use case).
I thought a Before/After-AllCallback was the best way to hook to the workflow, hence the extension.
And I thought the best way for that kind of extension to be not bound to annotations was to be enabled by default and then to check for system properties or annotations in order to eventually enable the feature.

Basically what is enabled by default is a _controller_, implemented by a JUnit extension, that enable the feature (part of the core) based on some conditions but please let me know whether you'd suggest something different for that part.

**Comment by fabiobrz** [comment]
> Could you please review the visibility of classes? It looks like all are public but are only used by extension within the package, perhaps they can be changed to package-private? Let's expose as minimum as possible.

Will do, thanks @simkam

File: junit5/src/main/resources/META-INF/services/org.junit.jupiter.api.extension.Extension (hunk 1 above)
Comment (simkam) [reply junit5/src/main/resources/META-INF/services/org.junit.jupiter.api.extension.Extension:1]: I understand, but it is different from other extensions do here in XTF. With other extensions, users need explicitly load them in their projects, no matter whether the feature is disabled or enabled.

Comment (fabiobrz) [reply junit5/src/main/resources/META-INF/services/org.junit.jupiter.api.extension.Extension:1]: Sure, I see. I'll think about the changes and hopefully come up with something soon. Thanks for your feedback.

**Comment by mnovak1** [comment]
@fabiobrz For me there was no issue to enable the feature and have colored text on standard output. I wanted to try how it will behave when forwarded to a file like `-Dxtf.sls.config="target=.*Test;output=/home/mnovak/sls-logs.txt"` but no `/home/mnovak/sls-logs.txt` file was created. It's still logged to standard output. Is there anything else needed to configure?

**Comment by fabiobrz** [comment]
> @fabiobrz For me there was no issue to enable the feature and have colored text on standard output. I wanted to try how it will behave when forwarded to a file like `-Dxtf.sls.config="target=.*Test;output=/home/mnovak/sls-logs.txt"` but no `/home/mnovak/sls-logs.txt` file was created. It's still logged to standard output. Is there anything else needed to configure?

Hi @mnovak1 - Actually `output` should point to a directory which is meant to be the base path in which files for each executed test class will be created.
There are (at least) a couple of issues, though:
1. the implementation should create such path if it does not exist (?)
2. it should be better documented that the expected path is about a directory rather than a file.

Could you please take such information into account and try again? I had my files generated and would appreciate a confirmation from you.
I'll fix what above as soon as I push something to implement the changes in the feature activation logic that have been previously requested.

**Comment by simkam** [comment]
wrt feature activation, I think that only removing it from `services` and documenting it should be enough. Just to keep it consistent with other extensions.

**Comment by fabiobrz** [comment]
> wrt feature activation, I think that only removing it from `services` and documenting it should be enough. Just to keep it consistent with other extensions.

Thanks @simkam - so basically according to your suggestion, the users should:

- register the extension, e.g. adding the FQN to `services` locally  
- use the system property or annotation to enable the feature when running tests

Correct?

**Comment by simkam** [comment]
Yes, exactly

**Comment by mnovak1** [comment]
> Hi @mnovak1 - Actually `output` should point to a directory which is meant to be the base path in which files for each executed test class will be created. There are (at least) a couple of issues, though:
> 
> 1. the implementation should create such path if it does not exist (?)
> 2. it should be better documented that the expected path is about a directory rather than a file.
> 
> Could you please take such information into account and try again? I had my files generated and would appreciate a confirmation from you. I'll fix what above as soon as I push something to implement the changes in the feature activation logic that have been previously requested.

Thanks, now it works. +1 for both points.

**Comment by simkam** [comment]
> About this, @simkam - actually the extension lives in a different module and package as such, hence by a quick check, only PodLogsWatcher could be package-private. Am I missing something about this specific request?

No, sorry, I didn't notice extension is in different package.

File: junit5/src/main/resources/META-INF/services/org.junit.jupiter.api.extension.Extension (hunk 1 above)
Comment (fabiobrz) [reply junit5/src/main/resources/META-INF/services/org.junit.jupiter.api.extension.Extension]: Extension registration removed, resolving this thread.

**Comment by fabiobrz** [comment]
Hi @simkam - I pushed some changes in order to address your comments. Feel free to let me know whether you've got any additional feedback.

**Review by simkam — APPROVED** [review]

**Comment by simkam** [comment]
LGTM

File: README.md [hunk 2]
```diff
@@ -157,6 +157,91 @@ xtf.foo.v2.version=1.0.3
...
+```java
+@ServiceLogsStreaming
+@Slf4j
+public class HelloWorldTest {
+  // ...
+}
+```
+
+###### The `xtf.sls.enabled` and `xtf.sls.config` property (Developer/Automation perspective)
+You can enable the SLS feature by setting the `xtf.sls.enabled` property so that it would apply to
```
Comment (mnovak1) [inline README.md:197]: Just a suggestion. `sls` seems to hard to remember (at least to me :-) ) Could it be renamed to `xtf.log.streaming.enabled` ? wdyt?

Comment (fabiobrz) [reply README.md:197]: Thanks @mnovak1, will do.
